### PR TITLE
ContestListView Enhancements on backend

### PR DESCRIPTION
**Pull request by sanyavertolet**
### What's done:
 * Added `featured_contest` table
 * Added several endpoints to manage `featured_contest` table
 * Added `creationTime` column into `contest` table
 * Added endpoint to get N newest contests

### CODE REVIEW

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 1]
```diff
@@ -70,9 +74,51 @@ internal class ContestController(
...
+        method = "POST",
+        summary = "Create or delete featured contest.",
+        description = "Mark contest to be featured if it is not marked so yet or unmark otherwise.",
+    )
+    @Parameters(
+        Parameter(name = "contestDto", `in` = ParameterIn.DEFAULT, description = "contest requested for creation", required = true),
+    )
+    @ApiResponse(responseCode = "200", description = "Contest was successfully created.")
+    @ApiResponse(responseCode = "404", description = "Contest with given name is not found.")
+    fun addOrDeleteContestToFeatured(
```
Comment (akuleshov7) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:93]: sorry if I am missing that, but can we check that superusers do this?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 2]
```diff
@@ -70,9 +74,51 @@ internal class ContestController(
...
+    @GetMapping("/featured/list")
+    @Operation(
+        method = "GET",
+        summary = "Get featured contests.",
+        description = "Get list of contests marked by admins as featured.",
+    )
+    @ApiResponse(responseCode = "200", description = "Contests were successfully fetched.")
+    fun getFeaturedContests(): Flux<ContestDto> = Flux.fromIterable(contestService.getFeaturedContests())
+        .filter {
+            LocalDateTime.now() < it.endTime
```
Comment (akuleshov7) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:116]: may be rename to list-active then?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ContestRepository.kt [hunk 3]
```diff
@@ -72,4 +77,20 @@ JpaSpecificationExecutor<Contest> {
      * @return [Page] of [Contest]s that belong to organization with name [organizationName]
      */
     fun findByOrganizationName(organizationName: String, pageable: Pageable): Page<Contest>
+
+    fun findByStatusNotAndEndTimeAfterOrderByCreationTimeDesc(status: ContestStatus, currentTime: LocalDateTime, pageable: Pageable): List<Contest>
+
+    @Query("""select contest_id from save_cloud.featured_contests""", nativeQuery = true)
```
Comment (akuleshov7) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ContestRepository.kt:83]: very simple query, why native and not JPA?

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ContestRepository.kt [hunk 4]
```diff
@@ -72,4 +77,20 @@ JpaSpecificationExecutor<Contest> {
...
+    fun findFeaturedContestIds(): Set<Long>
+
+    @Query("""select contest_id from save_cloud.featured_contests where contest_id = :id""", nativeQuery = true)
+    fun findFeaturedContestById(@Param("id") contestId: Long): Long?
+
+    @Modifying
+    @Query("""insert into save_cloud.featured_contests values (null, :id)""", nativeQuery = true)
```
Comment (akuleshov7) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ContestRepository.kt:90]: same

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt [hunk 5]
```diff
@@ -91,9 +99,29 @@ class ContestService(
     @Suppress("FUNCTION_BOOLEAN_PREFIX")
     fun createContestIfNotPresent(newContest: Contest): Boolean =
             if (contestRepository.findByName(newContest.name).isEmpty) {
-                contestRepository.save(newContest)
+                contestRepository.save(newContest.apply {  })
```
Comment (akuleshov7) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt:102]: what's for?

**Review by akuleshov7 — APPROVED** [review]

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt (hunk 1 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:93]: There is annotation for this on line 82:
```kotlin
@PreAuthorize("hasRole('ROLE_SUPER_ADMIN')")
```

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt (hunk 2 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:116]: Let it be so

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ContestRepository.kt (hunk 3 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/ContestRepository.kt:83]: As far as I understand, we are required to have an entity if we want to use JPQL while there is no need in entity for such a simple table

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt (hunk 5 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt:102]: Should be removed

Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ContestService.kt]: Done

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt (hunk 2 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:116]: Done

**Pull request merged**